sql/colexec: TestHashRouterRandom failed

[(sql/colexec).TestHashRouterRandom failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2580677&tab=buildLog) on [master@b9eb459b18369041f9e67c4830aa316e9d2ca0f0](https://github.com/cockroachdb/cockroach/commits/b9eb459b18369041f9e67c4830aa316e9d2ca0f0):

```
=== RUN   TestHashRouterRandom
    test_log_scope.go:72: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestHashRouterRandom996257729
    test_log_scope.go:73: use -show-logs to present logs inline
=== CONT  TestHashRouterRandom
    routers_test.go:1204: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestHashRouterRandom996257729
--- FAIL: TestHashRouterRandom (12.45s)
=== RUN   TestHashRouterRandom/terminationScenario=0/numOutputs=45/blockedThreshold=9093/totalInputSize=11180/hashCols=[0_1]/memoryLimit=37_MiB
    routers_test.go:1133: unexpected number of results 11077, expected 11180
    --- FAIL: TestHashRouterRandom/terminationScenario=0/numOutputs=45/blockedThreshold=9093/totalInputSize=11180/hashCols=[0_1]/memoryLimit=37_MiB (1.46s)
=== RUN   TestHashRouterRandom/terminationScenario=0/numOutputs=45/blockedThreshold=9093/totalInputSize=11180/hashCols=[0_1]/memoryLimit=1.0_GiB
    routers_test.go:1133: unexpected number of results 11178, expected 11180
    --- FAIL: TestHashRouterRandom/terminationScenario=0/numOutputs=45/blockedThreshold=9093/totalInputSize=11180/hashCols=[0_1]/memoryLimit=1.0_GiB (2.08s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestHashRouterRandom PKG=./pkg/sql/colexec TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestHashRouterRandom.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(sql/colexec).TestHashRouterRandom failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2581729&tab=buildLog) on [master@55d93310bc8f31c29435330617ca78be789899c5](https://github.com/cockroachdb/cockroach/commits/55d93310bc8f31c29435330617ca78be789899c5):

```
=== RUN   TestHashRouterRandom
    test_log_scope.go:72: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestHashRouterRandom800717486
    test_log_scope.go:73: use -show-logs to present logs inline
=== CONT  TestHashRouterRandom
    routers_test.go:1204: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestHashRouterRandom800717486
--- FAIL: TestHashRouterRandom (1.36s)
=== RUN   TestHashRouterRandom/terminationScenario=0/numOutputs=28/blockedThreshold=2997/totalInputSize=1284/hashCols=[0_1]/memoryLimit=8.2_MiB
    routers_test.go:1133: unexpected number of results 1273, expected 1284
    --- FAIL: TestHashRouterRandom/terminationScenario=0/numOutputs=28/blockedThreshold=2997/totalInputSize=1284/hashCols=[0_1]/memoryLimit=8.2_MiB (0.23s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestHashRouterRandom PKG=./pkg/sql/colexec TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestHashRouterRandom.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(sql/colexec).TestHashRouterRandom failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2582196&tab=buildLog) on [master@5bd34572e489dea0fa67feedd7c499f0ad9f9ebd](https://github.com/cockroachdb/cockroach/commits/5bd34572e489dea0fa67feedd7c499f0ad9f9ebd):

```
=== RUN   TestHashRouterRandom
    test_log_scope.go:72: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestHashRouterRandom232140531
    test_log_scope.go:73: use -show-logs to present logs inline
=== CONT  TestHashRouterRandom
    routers_test.go:1204: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestHashRouterRandom232140531
--- FAIL: TestHashRouterRandom (2.72s)
=== RUN   TestHashRouterRandom/terminationScenario=0/numOutputs=11/blockedThreshold=2421/totalInputSize=3231/hashCols=[0]/memoryLimit=1.0_GiB
    routers_test.go:1133: unexpected number of results 3229, expected 3231
    --- FAIL: TestHashRouterRandom/terminationScenario=0/numOutputs=11/blockedThreshold=2421/totalInputSize=3231/hashCols=[0]/memoryLimit=1.0_GiB (0.31s)
=== RUN   TestHashRouterRandom/terminationScenario=0/numOutputs=11/blockedThreshold=2421/totalInputSize=3231/hashCols=[0]/memoryLimit=31_MiB
    routers_test.go:1133: unexpected number of results 3211, expected 3231
    --- FAIL: TestHashRouterRandom/terminationScenario=0/numOutputs=11/blockedThreshold=2421/totalInputSize=3231/hashCols=[0]/memoryLimit=31_MiB (0.44s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestHashRouterRandom PKG=./pkg/sql/colexec TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestHashRouterRandom.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/sql/colexec/routers_test.go
+++ b/pkg/sql/colexec/routers_test.go
@@ -31,7 +31,6 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/types"
 	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/colcontainerutils"
-	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
 	"github.com/cockroachdb/cockroach/pkg/util/humanizeutil"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
@@ -861,7 +860,6 @@ func TestHashRouterOneOutput(t *testing.T) {
 
 func TestHashRouterRandom(t *testing.T) {
 	defer leaktest.AfterTest(t)()
-	skip.WithIssue(t, 58956, "flaky test")
 	defer log.Scope(t).Close(t)
 	ctx := context.Background()
 

--- a/pkg/sql/colexec/spilling_queue.go
+++ b/pkg/sql/colexec/spilling_queue.go
@@ -238,11 +238,11 @@ func (q *spillingQueue) enqueue(ctx context.Context, batch coldata.Batch) error 
 
 	alreadyCopied := 0
 	if q.numInMemoryItems > 0 {
-		// If we have already enqueued at least one batch, let's try to append
-		// as many tuples to it as it has the capacity for.
+		// If we have already enqueued at least one batch, let's try to copy
+		// as many tuples into it as it has the capacity for.
 		tailBatchIdx := q.curTailIdx - 1
 		if tailBatchIdx < 0 {
-			tailBatchIdx = 0
+			tailBatchIdx = len(q.items) - 1
 		}
 		tailBatch := q.items[tailBatchIdx]
 		if l, c := tailBatch.Length(), tailBatch.Capacity(); l < c {
@@ -252,13 +252,15 @@ func (q *spillingQueue) enqueue(ctx context.Context, batch coldata.Batch) error 
 			}
 			q.unlimitedAllocator.PerformOperation(tailBatch.ColVecs(), func() {
 				for i := range q.typs {
-					tailBatch.ColVec(i).Append(
-						coldata.SliceArgs{
-							Src:         batch.ColVec(i),
-							Sel:         batch.Selection(),
-							DestIdx:     l,
-							SrcStartIdx: 0,
-							SrcEndIdx:   alreadyCopied,
+					tailBatch.ColVec(i).Copy(
+						coldata.CopySliceArgs{
+							SliceArgs: coldata.SliceArgs{
+								Src:         batch.ColVec(i),
+								Sel:         batch.Selection(),
+								DestIdx:     l,
+								SrcStartIdx: 0,
+								SrcEndIdx:   alreadyCopied,
+							},
 						},
 					)
 				}

